rcl/src/rcl/publisher.c

```diff
@@ -37,6 +37,7 @@ typedef struct rcl_publisher_impl_t
   rcl_publisher_options_t options;
   rcl_context_t * context;
   rmw_publisher_t * rmw_handle;
+  rmw_publisher_allocation_t * allocation;
```

Comment:
You cannot just store a publish preallocation in the publisher because it is not thread safe. The preallocation needs to be created by the user or by rclcpp in order to ensure it is used in a thread safe manner.

Reply:
I agree. In that case the RCL should be changed or augmented to allow the user to pass the allocation to RCL.